File: 98.validate-binary-search-tree.go
package main

import "math"

/*
 * @lc app=leetcode id=98 lang=golang
 *
 * [98] Validate Binary Search Tree
 */

// @lc code=start
// *
// Definition for a binary tree node.
// type TreeNode struct {
// 	Val   int
// 	Left  *TreeNode
// 	Right *TreeNode
// }

func isValidBST(root *TreeNode) bool {
	return isValidbst(root, math.Inf(-1), math.Inf(1))
}

func isValidbst(root *TreeNode, min, max float64) bool {
	if root == nil {
		return true
	}
	v := float64(root.Val)
	return v < max && v > min && isValidbst(root.Left, min, v) && isValidbst(root.Right, v, max)
}

// @lc code=end

// class Solution {
//   public boolean isValidBST(TreeNode root) {
//     return isValidBST(root, null, null);
//   }

//   private boolean isValidBST(TreeNode root, Integer min, Integer max) {
//     if (root == null) return true;
//     if ((min != null && root.val <= min)
//       ||(max != null && root.val >= max))
//         return false;

//     return isValidBST(root.left, min, root.val)
//         && isValidBST(root.right, root.val, max);
//   }
// }

File: 98.validate-binary-search-tree.py
#
# @lc app=leetcode id=98 lang=python3
#
# [98] Validate Binary Search Tree
#

# @lc code=start
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None) -> None:  # type: ignore
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        self.prev: TreeNode = None  # type: ignore
        return self.inOrder(root)

    def inOrder(self, root: TreeNode) -> bool:
        if not root:
            return True
        if not self.inOrder(root.left):
            return False
        if self.prev and root.val <= self.prev.val:
            return False
        self.prev = root
        return self.inOrder(root.right)


# @lc code=end
